src/core/layertree/qgslayertreemodel.cpp

```diff
@@ -1471,8 +1471,33 @@ void QgsLayerTreeModel::addLegendToLayer( QgsLayerTreeLayer *nodeL )
 }
 
 
-QgsLayerTreeModel::LayerLegendTree *QgsLayerTreeModel::tryBuildLegendTree( const QList<QgsLayerTreeModelLegendNode *> &nodes )
+QgsLayerTreeModel::LayerLegendTree *QgsLayerTreeModel::tryBuildLegendTree( const QList<QgsLayerTreeModelLegendNode *> &nodes, Qgis::LayerType layerType )
 {
+  if ( layerType == Qgis::LayerType::Mesh )
+  {
+    QHash<QString, QgsLayerTreeModelLegendNode *> rule2node;
+    rule2node[QString()] = nullptr;
+    for ( QgsLayerTreeModelLegendNode *n : nodes )
+    {
+      QString key = n->objectName();
+      if ( key.isEmpty() ) // in tree all nodes must have key
+        return nullptr;
+      if ( rule2node.contains( key ) ) // and they must be unique
+        return nullptr;
+      rule2node[key] = n;
+    }
+
+    LayerLegendTree *tree = new LayerLegendTree;
+    for ( QgsLayerTreeModelLegendNode *n : nodes )
+    {
+      const QString parentName = n->property( "parentNode" ).toString();
+      QgsLayerTreeModelLegendNode *parent = rule2node.value( parentName, nullptr );
+      tree->parents[n] = parent;
+      tree->children[parent] << n;
+    }
+    return tree;
+  }
+
```

Comment:
Duplicating this code for mesh layers and relying on custom properties feels a little hacky...
I would suggest to leave this class as is and re-use existing infrastructure for tree-like legend items.
I think you should populate `QgsLayerTreeModelLegendNode::RuleKeyRole` for the `QgsSimpleLegendNode`s and `QgsLayerTreeModelLegendNode::ParentRuleKeyRole` for the `QgsRasterSymbolLegendNode`s and `QgsColorRampLegendNode`s (you'll have to update their `data`/`setData` methods as they do not support it now) and the tree should populate itself!

Reply:
Thanks for the advice. I'll  be pushing my commit in a bit. Btw, the KeyRoles are encapsulated in the classes and only settable upon construction and retrievable via the data method, so I stuck to those restrictions.